backend/src/api/bitcoin/bitcoin.routes.ts

```diff
@@ -241,6 +243,74 @@ class BitcoinRoutes {
     }
   }
 
+  /**
+   * Takes the PSBT as text/plain body, parses it, and adds the full
+   * parent transaction to each input that doesn't already have it.
+   * This is used for BTCPayServer / Trezor users which need access to
+   * the full parent transaction even with segwit inputs.
+   * It will respond with a text/plain PSBT in the same format (hex|base64).
+   */
+  private async postPsbtCompletion(req: Request, res: Response) {
+    res.setHeader('content-type', 'text/plain');
+    const notFoundError = `Couldn't get transaction hex for parent of input`;
+    try {
+      let psbt: bitcoinjs.Psbt;
+      let format: 'hex' | 'base64';
+      let isModified = false;
+      try {
+        psbt = bitcoinjs.Psbt.fromBase64(req.body);
+        format = 'base64';
+      } catch (e1) {
+        try {
+          psbt = bitcoinjs.Psbt.fromHex(req.body);
+          format = 'hex';
+        } catch (e2) {
+          throw new Error(`Unable to parse PSBT`);
+        }
+      }
+      for (const [index, input] of psbt.data.inputs.entries()) {
+        if (!input.nonWitnessUtxo) {
+          // Buffer.from ensures it won't be modified in place by reverse()
+          const txid = Buffer.from(psbt.txInputs[index].hash)
+            .reverse()
+            .toString('hex');
+
+          let transaction: IEsploraApi.Transaction;
+          // If missing transaction, return 404 status error
+          try {
+            transaction = await bitcoinApi.$getRawTransaction(txid, true);
+            if (!transaction.hex) {
+              throw new Error('');
+            }
+          } catch (err) {
+            throw new Error(`${notFoundError} #${index} @ ${txid}`);
+          }
+
+          psbt.updateInput(index, {
+            nonWitnessUtxo: Buffer.from(transaction.hex, 'hex'),
+          });
```

Comment:
according to bip 174, for witness inputs the witnessUtxo must be filled in, and for legacy inputs the nonWitnessUtxo, but also allows witness inputs to have nonWitnessUtxo.

when i `walletprocesspsbt` a psbt with nonWitnessUtxo and no witnessUtxo for a witness input (from electrum), bitcoind fills in the witnessUtxo, and strips the input witness from nonWitnessUtxo. This way wallets that need witnessUtxo can read it, wallets that dont understand segwit dont encounter witness data in nonWitnessUtxo and wallets that check the txid can still do this because witness data does not affect the txid. i thinks it is best to stay in line with bitcoind's implementation

Reply:
Thoughts @NicolasDorier ?

This feature is a request from BTCPayServer, so I delegate this design decision to them.

Reply:
I didn't know that some implementation were removing the witness from nonWitnessUTXO. I never did it and never got problems, I guess this is because any wallet sufficiently advanced to know about PSBT already know what is witness data. I would prefer you don't strip the witness it as I don't have experience on what happen on hardware wallets. (btw, having both witnessUTXO and nonWitnessUTXO wasn't supported by some wallet at one point, but I guess it's ok now)
I also prefer not stripping witness data, because I plan to save the transaction in our cache, and I don't expect not having the full tx.

An idea if you can't get your mind on that would be to allow query parameters to the URL to control this, so on BTCPay side we would be able to redirect users on exactly the right way to do the PSBT. (for example, if we know a wallet don't support both witnessUTXO and nonWitnessUTXO we could redirect on the form where witnessUTXO isn't filled)

But this is maybe too much for this PR at this point, better to experiment with it and act on the feedback.